Exporting a Collection's metadata, permissions, and members

cam156 opened the pull request:
Fixes #2155 

Exports a Collection to json format including metadata, permissions and member ids (the members are expected to be exported separately see #2268).

The intention of this class is to act as a way to transfer data from Sufia 6 so that it can be imported into Sufia 7. See #2136 for user story.

I pulled much of the code from a branch in ScholarSphere written by @hectorcorrea export_s6

@projecthydra/sufia-code-reviewers

mjgiarlo commented on lib/sufia/export/converter.rb:
```diff
@@ -0,0 +1,21 @@
+      # @param [Boolean] pretty pass true to output formatted json using pretty_generate
+      def to_json(pretty = false)
+        return super unless pretty
+        JSON.pretty_generate(JSON.parse(to_json))
```
Is this method recursive? Or should this `to_json` be `super`?

hectorcorrea replied:
Good catch. I guess it never got into an infinite loop because the guard class the the top but I think it should be super.

cam156 replied:
@mjgiarlo & @hectorcorrea I tried super and that was no good (which is strange) calling to_json without the true then calls super.  Do not ask me why that is different from calling to_json, but the tests fail when I tried the change with the last PR.

mjgiarlo replied:
:eyes:  Huh.

cam156 replied:
@mjgiarlo I fooled around with some different ways to call super and `super()` works `super` does not it gives a 
```
     TypeError:
       can't dup TrueClass
```
The result is the same either way.  Which is easier to understand? recalling to_json with no parameters or calling super()?

hectorcorrea replied:
I would go for `super()` to remove the recursive call.

I think `super` is a keyword whereas `super()` is a function call. Ruby might not like the keyword inside another function call.

mjgiarlo commented on spec/lib/sufia/export/collection_converter_spec.rb:
```diff
@@ -0,0 +1,32 @@
+
+describe Sufia::Export::CollectionConverter do
+  let(:collection) do
+    Collection.create(title: "title1", creator: ["creator1"], description: "description1") do |col|
```
Should `title` be an array here?

mjgiarlo replied:
Yes, I do believe they should, according to https://github.com/projecthydra/sufia/blob/master/spec/factories/collections_factory.rb#L23-L26

Would it be possible to use a factory here instead of creating a repository object? That'd help with suite speed.

cam156 replied:
@mjgiarlo This is sufia 6, so there are no factories in existence. Also for Sufia 6 title and description are singular.

mjgiarlo replied:
Ah, right! This context-switching is twisting my wig. Thanks, @cam156.

mjgiarlo commented on lib/sufia/export/converter.rb:
```diff
@@ -0,0 +1,21 @@
+      # @param [Boolean] pretty pass true to output formatted json using pretty_generate
+      def to_json(pretty = false)
+        return super unless pretty
+        JSON.pretty_generate(JSON.parse(super()))
```
Thoughts on the `super()` here, @projecthydra/sufia-code-reviewers @jcoyne @jeremyf etc.

cam156 replied:
The other choice is to call to_json again with no parameter or false...
```
JSON.pretty_generate(JSON.parse(to_json))
```
or
```
JSON.pretty_generate(JSON.parse(to_json(false)))
```

cam156 replied:
or
```
json = super()
return json unless pretty
JSON.pretty_generate(JSON.parse(json))
```

cam156 commented on lib/sufia/export/converter.rb:
```diff
@@ -0,0 +1,22 @@
+      def to_json(pretty = false)
+        json = super()
+        return json unless pretty
+        JSON.pretty_generate(JSON.parse(json))
```
@mjgiarlo Went with a third option suggested by @awead and approved as readable by @mtribone. 👌 ?

cam156 commented:
@mjgiarlo, @awead & @hectorcorrea  I changed the method to match the calling parameters of the original to_json which allows for an options hash.  I moved pretty so that it would be an option.  This allows us to call super without the parens.  Hopefully the new solution ticks all the requests...

mjgiarlo commented:
Taking a look, @cam156!

mjgiarlo commented on lib/sufia/export/converter.rb:
```diff
@@ -0,0 +1,23 @@
+      # overrides to_json to optionally allow for a pretty version of the json to be outputted
+      #
+      # @param [Boolean] pretty pass true to output formatted json using pretty_generate
+      def to_json(options = nil)
```
Why set `options` to `nil` here? Then you have to check if it's `#blank?` on the next line. You could make this less ask-y and more tell-y by doing this:

```ruby
def to_json(options = {})
  json = super
  return json unless options[:pretty]
  JSON.pretty_generate(JSON.parse(json))
end
```

I don't care strongly about this, though. Are you inclined to care about that, or :shipit:?

cam156 replied:
@mjgiarlo I can make the change.

mjgiarlo commented:
@cam156 I have just one comment, but like I say, I don't care very much about it. Should I :shipit: or do you want to make that one change?

mjgiarlo commented:
TRAVIIIIIIIIIIIIIIIIIIS
